src/metabase/lib/computed.cljc

```diff
@@ -0,0 +1,92 @@
+
+  Only a corner of the vision is implemented here so far, it's expanding as needed in response to user perf issues."
+  (:require
+   [metabase.util :as u]
+   [metabase.util.log :as log]))
+
+(def ^:private weak-map
+  "A global weak map using queries as keys. Since CLJ(S) maps are immutable, if the map changes it is a new pointer,
+  and hence a new cache key. Since the keys of a weak map are weakly held, the map entries will be GC'd when the query
+  is, avoiding any extra memory retention."
+  #?(:clj  (java.util.Collections/synchronizedMap (java.util.WeakHashMap.))
+     :cljs (js/WeakMap.)))
+
+(def ^:dynamic *computed-cache*
+  "Dynamic var that holds an atom used for caching derived values for [[with-cache-sticky*]].
+
+  `nil` by default, meaning would-be sticky values are actually cached in the weak map, keyed by query. That's more
+  ephemeral than they're declared to be, but it's better than nothing.
+
+  It is always safe, in terms of correctness, to bind this to an `(atom {})` before making lib calls.
+  However, there is no LRU or other eviction logic, and so the memory use will continue to grow as it is held.
+  The best approach is to wrap a chunk of related lib calls (eg. `qp.preprocess`) with binding this var.
+
+  Soon this limitation should be removed by making queries `map`-likes with a private atom and evict-on-update."
+  nil)
+
+(defn- weak-atoms [query]
+  #?(:clj  (.computeIfAbsent ^java.util.WeakHashMap weak-map query
+                             (fn [_]
+                               {:sticky (atom {})
+                                :weak   (atom {})}))
+     :cljs (if (.has weak-map query)
+             (.get weak-map query)
+             (let [m {:sticky (atom {})
+                      :weak   (atom {})}]
+               (.set weak-map query m)
+               m))))
+
+(defn ^:dynamic *cache-hit-hook*
+  "A function `(f cache-key)` called whenever the sticky or ephemeral caches are hit.
+
+  Can be overridden for testing. The default just does a [[log/debug]]."
+  [cache-key]
+  (log/debug (str (u/colorize :green "HIT: ") (name (first cache-key)) " " (hash (rest cache-key)))))
+
+(defn ^:dynamic *cache-miss-hook*
+  "A function `(f cache-key)` called whenever the sticky or ephemeral caches are missed.
+
+  Can be overridden for testing. The default just does a [[log/debug]]."
+  [cache-key]
+  (log/debug (str (u/colorize :red "MISS: ") (name (first cache-key)) " " (hash (rest cache-key)))))
+
+(defn- cache-through [*atom cache-key f]
+  (let [output (get-in @*atom cache-key ::not-found)]
+    (if (not= output ::not-found)
+      (do (*cache-hit-hook* cache-key)
+          output)
+      (let [output (f)]
+        (swap! *atom assoc-in cache-key output)
+        (*cache-miss-hook* cache-key)
+        output))))
+
+(defn with-cache-none*
+  "A dummy cache function useful for checking if something is cached when it shouldn't be.
```

Comment:
Can you elaborate on how it is to be used?

Reply:
Done.